src/chrome/content/ff-overlay.js

```diff
@@ -85,30 +85,37 @@ if (typeof SL_Overlay === "undefined") {
                 .addEventListener("popupshowing", (event) => {
                     this.showFirefoxContextMenu(event);
                 }, false);
-            DownloadManager.connectToNas();
         };
 
         this.onMenuItemLinkCommand = (event) => {
-            if (DownloadManager.isConnected) {
-                window.open(DownloadManager.urlToConnect + "/webman/index.cgi?launchApp=SYNO.SDS.DownloadStation.Application", "Diskstation", this.strWindowFeatures);
-            } else {
-                Notification.show("No Connection");
-            }
+            DownloadManager.connectAndRun(() => {
+                if (DownloadManager.isConnected) {
+                    window.open(DownloadManager.urlToConnect + "/webman/index.cgi?launchApp=SYNO.SDS.DownloadStation.Application", "Diskstation", this.strWindowFeatures);
+                } else {
+                    Notification.show("No Connection");
+                }
+            });
         };
 
         this.onMenuItemCommand = (event) => {
-            if (DownloadManager.isConnected) {
-                DownloadManager.transferToNas(gContextMenu.linkURL);
-            } else {
-                Notification.show("No Connection");
-            }
+            let linkURL = gContextMenu.linkURL;
+            DownloadManager.connectAndRun(() => {
```

Comment:
I see what you mean. The only reason I don't *like* that, is because it isn't DRY. So for each function / method that needs a valid connection before executing, will contain the same or very similar code, making it more prone to error and not so pretty to manage.

I'll let you do this if that's ok, so I can learn and see how it's done the way you suggest. Maybe I'm overthinking and have a wrong idea in my head, not sure.